File: javascript/genome.js
/**
 * Genome class.
 * @class
 * @constructor
 * @public
 */
class Genome {
	/**
	 *
	 * @param {number} input number of input nodes
	 * @param {number} output number of output nodes
	 * @param {number} id id of the genome
	 * @param {boolean} offSpring ???
	 */
	constructor(input, output, id, offSpring = false) {
		// Genome parameters
		this.inputs = input;
		this.outputs = output;
		this.id = id;
		this.layers = 2;
		this.nextNode = 0;

		this.nodes = [];
		this.connections = [];

		if (!offSpring) {
			for (let i = 0; i < input; i++) {
				this.nodes.push(new Node(this.nextNode, 0));
				this.nextNode++;
			}

			for (let i = 0; i < output; i++) {
				this.nodes.push(new Node(this.nextNode, 1, true));
				this.nextNode++;
			}

			for (let i = 0; i < input; i++) {
				for (let j = input; j < output + input; j++) {
					const weight =
						Math.random() * this.inputs * Math.sqrt(2 / this.inputs);
					this.connections.push(
						new Connection(this.nodes[i], this.nodes[j], weight)
					);
				}
			}
		}
	}

	// Network Core
	generateNetwork() {
		// Clear all output Connection in the nodes
		const nodesLen = this.nodes.length;
		for (let i = 0; i < nodesLen; i++) {
			const node = this.nodes[i];
			node.outputConnections.splice(0, node.outputConnections.length);
		}
		// add the connections to the nodes
		const connectionsLen = this.connections.length;
		for (let i = 0; i < connectionsLen; i++) {
			const connection = this.connections[i];
			connection.fromNode.outputConnections.push(connection);
		}
		// prepare for feed forward
		this.sortByLayer();
	}

	/**
	 *
	 * @param {number[]} inputs array of inputs
	 * @returns {number[]} array of outputs
	 */
	feedForward(inputs) {
		this.generateNetwork(); // connect all the nodes

		// clear old inputs
		const nodesLen = this.nodes.length;
		for (let i = 0; i < nodesLen; i++) {
			this.nodes[i].inputSum = 0;
		}
		//assign new inputs
		for (let i = 0; i < this.inputs; i++) {
			this.nodes[i].outputValue = inputs[i];
		}
		// engage all nodes and Extract the results from the output nodes
		const result = [];
		for (let i = 0; i < nodesLen; i++) {
			const node = this.nodes[i];
			node.engage();
			if (node.output) {
				result.push(node.outputValue);
			}
		}
		return result;
	}

	/**
	 * crossover
	 * @param {Genome} partner
	 *
	 * @returns {Genome} offspring genome
	 */
	crossover(partner) {
		//TODO unique id
		const offSpring = new Genome(this.inputs, this.outputs, 0, true); // Child genome
		offSpring.nextNode = this.nextNode;

		// Take all nodes from this parent - output node activation 50% - 50%
		const nodesLen = this.nodes.length;
		for (let i = 0; i < nodesLen; i++) {
			const node = this.nodes[i].clone();
			if (node.output) {
				const partnerNode = partner.nodes[partner.getNode(node.number)];
				if (Math.random() < 0.5) {
					node.activationFunction = partnerNode.activationFunction;
					node.bias = partnerNode.bias;
				}
			}
			offSpring.nodes.push(node);
		}

		// Randomly take connections form this or the partner network
		//? never used
		const maxLayer = 0;
		const connectionLen = this.connections.length;
		for (let i = 0; i < connectionLen; i++) {
			const index = this.commonConnection(
				this.connections[i].getInnovationNumber(),
				partner.connections
			);

			if (index != -1) {
				// have common connections

				const conn =
					Math.random() < 0.5
						? this.connections[i].clone()
						: partner.connections[index].clone();

				// Reassign nodes
				const fromNode =
					offSpring.nodes[offSpring.getNode(conn.fromNode.number)];
				const toNode = offSpring.nodes[offSpring.getNode(conn.toNode.number)];

				conn.fromNode = fromNode;
				conn.toNode = toNode;

				// Add this connection to the child
				if (fromNode && toNode) {
					offSpring.connections.push(conn);
				}
			} else {
				// no common connection => take from this
				const conn = this.connections[i].clone();

				// Reassign nodes
				const fromNode =
					offSpring.nodes[offSpring.getNode(conn.fromNode.number)];
				const toNode = offSpring.nodes[offSpring.getNode(conn.toNode.number)];

				conn.fromNode = fromNode;
				conn.toNode = toNode;

				// add this connection to the child
				if (fromNode && toNode) {
					offSpring.connections.push(conn);
				}
			}
		}
		offSpring.layers = this.layers;
		return offSpring;
	}

	// Mutation
	mutate() {
		// let mut;
		//? not exclusive event?

		if (Math.random() < 0.8) {
			// modify connection
			// mut = 'ModConn';
			for (let i = 0; i < this.connections.length; i++) {
				this.connections[i].mutateWeight();
			}
		}
		if (Math.random() < 0.5) {
			// modify bias
			// mut = 'ModBias';
			for (let i = 0; i < this.nodes.length; i++) {
				this.nodes[i].mutateBias();
			}
		}
		if (Math.random() < 0.1) {
			// modify activation function
			// mut = 'ModAct';
			let i = Math.floor(Math.random() * this.nodes.length);
			this.nodes[i].mutateActivation();
		}
		if (Math.random() < 0.01) {
			// add mode
			// mut = 'AddNode';
			this.addNode();
		}
	}

	addNode() {
		// add node to network
		// get a random connection to replace with a node
		let connectionIndex = Math.floor(Math.random() * this.connections.length);
		let pickedConnection = this.connections[connectionIndex];
		pickedConnection.enabled = false;
		// delete connection
		this.connections.splice(connectionIndex, 1);

		// create new node
		let newNode = new Node(this.nextNode, pickedConnection.fromNode.layer + 1);
		for (let i = 0; i < this.nodes.length; i++) {
			const node = this.nodes[i];
			if (node.layer > pickedConnection.fromNode.layer) {
				node.layer++;
			}
		}

		// new connections
		let newConnection1 = new Connection(pickedConnection.fromNode, newNode, 1); // set weight as 1
		let newConnection2 = new Connection(
			newNode,
			pickedConnection.toNode,
			pickedConnection.weight
		);
		this.layers++;
		this.connections.push(newConnection1);
		this.connections.push(newConnection2);
		this.nodes.push(newNode);
		this.nextNode++;
	}

	addConnection() {
		if (this.fullyConnected()) {
			return;
		}

		// choose nodes to connect
		let node1Index = Math.floor(Math.random() * this.nodes.length);
		let node2Index = Math.floor(Math.random() * this.nodes.length);
		// ensure nodes are valid, not the same and not connected
		//TODO factorize
		while (
			this.nodes[node1Index].layer == this.nodes[node2Index].layer ||
			this.nodesConnected(this.nodes[node1Index], this.nodes[node2Index])
		) {
			node1Index = Math.floor(Math.random() * this.nodes.length);
			node2Index = Math.floor(Math.random() * this.nodes.length);
		}

		// switch nodes based on their layers, ensure node1 is always on the left
		if (this.nodes[node1Index].layer > this.nodes[node2Index].layer) {
			let temp = node1Index;
			node1Index = node2Index;
			node2Index = temp;
		}

		// add connection
		let newConnection = new Connection(
			this.nodes[node1Index],
			this.nodes[node2Index],
			Math.random() * this.inputs * Math.sqrt(2 / this.inputs)
		);
		this.connections.push(newConnection);
	}

	// Utilities

	// search through all connections to check for one with the correct Innovation number
	commonConnection(innN, connections) {
		for (let i = 0; i < connections.length; i++) {
			if (innN == connections[i].getInnovationNumber()) {
				return i;
			}
		}
		// not found
		return -1;
	}

	// search if there is a connection between node1 and node2
	nodesConnected(node1, node2) {
		for (let i = 0; i < this.connections.length; i++) {
			const conn = this.connections[i];
			if (
				(conn.fromNode == node1 && conn.toNode == node2) ||
				(conn.fromNode == node2 && conn.toNode == node1)
			) {
				return true;
			}
		}
		return false;
	}

	// check if the network is fully connected
	fullyConnected() {
		let maxConnections = 0;
		let nodesPerLayer = [];

		// Calculate all possible connections
		for (let i = 0; i < this.nodes.length; i++) {
			const node = this.nodes[i];
			if (nodesPerLayer[node.layer] != undefined) {
				nodesPerLayer[node.layer]++;
			} else {
				nodesPerLayer[node.layer] = 1;
			}
		}
		for (let i = 0; i < this.layers - 1; i++) {
			for (let j = i + 1; j < this.layers; j++) {
				maxConnections += nodesPerLayer[i] * nodesPerLayer[j];
			}
		}
		// compare
		return maxConnections == this.connections.length;
	}
	// sort all nodes by layer
	sortByLayer() {
		this.nodes.sort((a, b) => a.layer - b.layer);
	}
	// return a copy of this genome
	clone() {
		let clone = new Genome(this.inputs, this.outputs, this.id);
		clone.nodes = this.nodes.slice(0, this.nodes.length);
		clone.connections = this.connections.slice(0, this.connections.length);
		return clone;
	}
	// returns the index of a node with that number
	getNode(x) {
		for (let i = 0; i < this.nodes.length; i++) {
			if (this.nodes[i].number == x) {
				return i;
			}
		}
		return -1;
	}

	// computational weight of the network
	calculateWeight() {
		return this.connections.length + this.nodes.length;
	}

	draw(width = 400, height = 400, container = 'svgContainer') {
		//Draw the genome to a svg
		var element = document.getElementById(this.id);
		if (element) element.parentNode.removeChild(element);

		var svg = d3
			.select('body')
			.append('svg')
			.attr('width', width)
			.attr('height', height)
			.attr('id', this.id);

		var force = d3.layout
			.force()
			.gravity(0.05)
			.distance(100)
			.charge(-100)
			.size([width, height]);

		let connections = [];
		this.connections.forEach((conn) => {
			connections.push({
				source: this.getNode(conn.fromNode.number),
				target: this.getNode(conn.toNode.number),
				weight: conn.weight,
				enabled: conn.enabled,
			});
		});

		let nodes = [];
		this.nodes.forEach((originalNode) => {
			let node = originalNode.clone();
			if (node.layer == 0) {
				node.fixed = true;
				node.y = height - height * 0.2;
				node.x = (width / this.inputs) * node.number + width / this.inputs / 2;
			}

			if (node.output) {
				node.fixed = true;
				node.y = height * 0.2;
				node.x =
					(width / this.outputs) * (node.number - this.inputs) +
					width / this.outputs / 2;
			}

			nodes.push(node);
		});

		force.nodes(nodes).links(connections).start();

		var link = svg
			.selectAll('.link')
			.data(connections)
			.enter()
			.append('line')
			.attr('class', 'link')
			.style('stroke-width', function (d) {
				return d.enabled
					? d.weight > 0
						? 0.3 + d.weight
						: 0.3 + d.weight * -1
					: 0;
			})
			.style('stroke', function (d) {
				return d.weight > 0 ? '#0f0' : '#f00';
			});

		var node = svg
			.selectAll('.node')
			.data(nodes)
			.enter()
			.append('g')
			.attr('class', 'node')
			.call(force.drag);

		node
			.append('circle')
			.attr('r', '5')
			.attr('fill', function (d) {
				return d.layer == 0 ? '#00f' : d.output ? '#f00' : '#000';
			});

		node
			.append('text')
			.attr('dx', 12)
			.attr('dy', '.35em')
			.text(function (d) {
				return (
					d.number +
					(d.layer > 0
						? '(' + activationsNames[d.activationFunction] + ')'
						: null)
				);
			});

		force.on('tick', function () {
			link
				.attr('x1', function (d) {
					return d.source.x;
				})
				.attr('y1', function (d) {
					return d.source.y;
				})
				.attr('x2', function (d) {
					return d.target.x;
				})
				.attr('y2', function (d) {
					return d.target.y;
				});

			node.attr('transform', function (d) {
				return 'translate(' + d.x + ',' + d.y + ')';
			});
		});

		var element = document.getElementById(this.id);
		document.getElementById(container).append(element);
	}
}

File: cpp/genome.cpp
#include "genome.h"
#include "node.h"
#include "connection.h"
#include "randomFun.h"
#include <cmath>

Genome::Genome(long int _input, long int _output, long int _id, bool _offSpring = false) {
    inputs = _input;
    outputs = _output;
    id = _id;
    layers = 2;
    nextNode = 0;

    connections = std::vector<Connection*>();
    nodes = std::vector<Node*>();


    if (!_offSpring) {
        for (long int i = 0; i < _input; ++i) {
            nodes.push_back(new Node(nextNode, 0));
            nextNode++;
        }

        for (long int i = 0; i < _output; ++i) {
            nodes.push_back(new Node(nextNode, 1, true));
            nextNode++;
        }

        for (long int i = 0; i < _input; ++i) {
            for (long int j = 0; j < _output + _input; ++j) {
                const long double weight = randNum(0, 1) * _input * sqrt(2.0 / (double)_input);
                connections.push_back(new Connection(nodes[i], nodes[j], weight));
            }
        }
    }
}

Genome::~Genome() {
    // delete node first to prevent double deallocation;
    const size_t sizeOfNodes = nodes.size();
    for (int i = 0; i < sizeOfNodes; ++i) {
        delete nodes[i];
        nodes[i] = nullptr;
    }
    const size_t sizeOfConnection = connections.size();
    for (int i = 0; i < sizeOfConnection; ++i) {
        delete connections[i];
        connections[i] = nullptr;
    }
}

void Genome::generateNetwork() {
    // clear all output connection in the nodes
    const long int nodesLen = nodes.size();
    for (long int i = 0; i < nodesLen;++i) {
        nodes[i]->outputConnections.clear();
    }
    // add the connection to the nodes
    const long int connectionsLen = connections.size();
    for (long int i = 0; i < nodesLen;i++) {
        Connection* connection = connections[i];
        connection->getFromNode()->outputConnections.push_back(connection);
    }
    // prepare for feed forward
    sortByLayer();
}

std::vector<long double> Genome::feedForward(const std::vector<long double>& input) {
    const long int nodesLen = nodes.size();
    for (long int i = 0; i < nodesLen; ++i) {
        nodes[i]->inputSum = 0;
    }
    // assign new inputs
    const long int inputLen = input.size();
    for (long int i = 0; i < inputLen; ++i) {
        nodes[i]->outputValue = input[i];
    }

    // engage all nodes and Extract the results from the output nodes
    std::vector<long double> result;
    for (long int i = 0; i < nodesLen; ++i) {
        Node* node = nodes[i];
        node->engage();
        if (node->output) {
            result.push_back(node->outputValue);
        }
    }

    return result;
}

Genome* Genome::crossover(const Genome* const partner) {

    Genome* offSpring = new Genome(inputs, outputs, 0, true);
    offSpring->nextNode = nextNode;

    // take all nodes from this parent - output node activation 50% - 50%
    const long int nodesLen = nodes.size();
    for (long int i = 0; i < nodesLen; ++i) {
        Node* node = nodes[i]->clone();
        if (node->output) {
            Node* partnerNode = partner->nodes[partner->getNode(node->number)];
            if (randNum(0, 1) < 0.5) {
                node->activationFunctionIndex = partnerNode->activationFunctionIndex;
                node->bias = partnerNode->bias;
            }
        }
        offSpring->nodes.push_back(node);
    }

    // Randomly take connections form this or the partner network
    //? never used
    const long int maxLayer = 0;
    const long int connectionLen = connections.size();
    for (long int i = 0; i < connectionLen; ++i) {
        const long int index = commonConnection(connections[i]->getInnovationNumber(), partner->connections);

        if (index != -1) {
            // have common connections

            Connection* connection = randNum(0, 1) < 0.5 ? connections[i]->clone() : partner->connections[index]->clone();

            // reassign nodes
            long int fromNodeIndex = offSpring->getNode(connection->getFromNode()->number);
            long int toNodeIndex = offSpring->getNode(connection->getToNode()->number);
            if (fromNodeIndex != -1 && toNodeIndex != -1) {

                Node* fromNode = offSpring->nodes[fromNodeIndex];
                Node* toNode = offSpring->nodes[toNodeIndex];

                connection->setFromNode(fromNode);
                connection->setToNode(toNode);

                offSpring->connections.push_back(connection);
            }

        } else {
            // no common connection => take from this
            Connection* connection = connections[i]->clone();

            long int fromNodeIndex = offSpring->getNode(connection->getFromNode()->number);
            long int toNodeIndex = offSpring->getNode(connection->getToNode()->number);

            if (fromNodeIndex != -1 && toNodeIndex != -1) {

                Node* fromNode = offSpring->nodes[fromNodeIndex];
                Node* toNode = offSpring->nodes[toNodeIndex];

                connection->setFromNode(fromNode);
                connection->setToNode(toNode);

                offSpring->connections.push_back(connection);
            }
        }
    }

    offSpring->layers = layers;
    return offSpring;
}

void Genome::mutate() {
    const long int connectionLen = connections.size();
    const long int nodesLen = nodes.size();
    if (randNum(0, 1) < 0.8) {
        for (long int i = 0; i < connectionLen; ++i) {
            connections[i]->mutateWeight();
        }
    }
    if (randNum(0, 1) < 0.5) {
        for (long int i = 0; i < nodesLen; ++i) {
            nodes[i]->mutateBias();
        }
    }
    if (randNum(0, 1) < 0.1) {
        const long int i = floor(randNum(0, nodesLen));
        nodes[i]->mutateActivationFunction();
    }
    if (randNum(0, 1) < 0.01) {
        addNode();
    }
}

void Genome::addNode() {
    // add node to network
    // get a random connection to replace with a node
    const long int connectionLen = connections.size();
    const long int connectionIndex = floor(randNum(0, connectionLen));
    const long int nodesLen = nodes.size();
    Connection* pickedConnection = connections[connectionIndex];
    // delete connection
    connections.erase(connections.begin() + connectionIndex);

    // create new node
    Node* newNode = new Node(nextNode, pickedConnection->getFromNode()->layer + 1);
    for (long int i = 0; i < nodesLen;++i) {
        if (nodes[i]->layer > pickedConnection->getFromNode()->layer) {
            ++nodes[i]->layer;
        }
    }

    // new connections
    Connection* newConnection1 = new Connection(pickedConnection->getFromNode(), newNode, 1);
    Connection* newConnection2 = new Connection(newNode, pickedConnection->getToNode(), pickedConnection->getWeight());

    ++layers;
    connections.push_back(newConnection1);
    connections.push_back(newConnection2);
    nodes.push_back(newNode);
    ++nextNode;
}

void Genome::addConnection() {
    if (fullyConnected()) return;
    const long int nodesLen = nodes.size();
    // choose nodes to connect
    long int node1Index = floor(randNum(0, nodesLen));
    long int node2Index = floor(randNum(0, nodesLen));
    //TODO factorize
    while (nodes[node1Index]->layer == nodes[node2Index]->layer || nodesConnected(nodes[node1Index], nodes[node2Index])) {
        node1Index = floor(randNum(0, nodesLen));
        node2Index = floor(randNum(0, nodesLen));
    }
    // switch nodes based on their layers, ensure node1 is on the left
    if (nodes[node1Index]->layer > nodes[node2Index]->layer) {
        long int tmp = node1Index;
        node1Index = node2Index;
        node2Index = tmp;
    }

    // add connection
    Connection* newConnection = new Connection(nodes[node1Index], nodes[node2Index], randNum(0, 1) * inputs * sqrt(2 / inputs));
    connections.push_back(newConnection);
}

long int Genome::commonConnection(long int innovationNumber, const std::vector<Connection* >& connections) const {

    const long int connectionLen = connections.size();
    for (long int i = 0; i < connectionLen; ++i) {
        if (connections[i]->getInnovationNumber() == innovationNumber) {
            return i;
        }
    }
    return -1;
}

bool Genome::nodesConnected(Node* node1, Node* node2) const {
    const long int connectionLen = connections.size();
    for (long int i = 0; i < connectionLen; ++i) {
        const Connection* conn = connections[i];
        if (
            ((conn->getFromNode()) == node1 && (conn->getToNode()) == node2) ||
            ((conn->getFromNode()) == node2 && (conn->getToNode()) == node1)
            ) {
            return true;
        }
    }
    return false;
}

bool Genome::fullyConnected() const { }

void Genome::sortByLayer() { }

Genome* Genome::clone() const { }

long int getNode(long int index) { }

long double Genome::calculateWeight() const { }






